### ISSUE TITLE
Installing reflex-0.5.0 fails with "Kind incompatibility when matching types" in Patch.hs

### ISSUE THREAD

**Issue by hansroland** [reflex-frp/reflex#99]
GHCJS Version: 0.2.0 (GHC 7.10.3).
GHCJS Download date: 2017-03-27.
I deleted directory /home/roland/.ghcjs and re-installed the GHCJS compiler.
Clone date of https://github.com/reflex-frp/reflex: 2017-03-27.

File  [reflex-ghcjs.txt](https://github.com/reflex-frp/reflex/files/872397/reflex-ghcjs.txt) contains the full installation log.

**Error:**

src/Reflex/Patch.hs:106:45:
    Kind incompatibility when matching types:
      g0 :: k -> *
      Constant () :: * -> *
    Expected type: k1 v1 -> ComposeMaybe v v1 -> Maybe (g0 v1)
      Actual type: k1 v1 -> ComposeMaybe v v1 -> Maybe (Constant () b0)
    Relevant bindings include
      deletions :: DMap k1 g0 (bound at src/Reflex/Patch.hs:106:11)
      old :: PatchTarget (PatchDMap k1 v)
        (bound at src/Reflex/Patch.hs:104:26)
      diff :: DMap k1 (ComposeMaybe v)
        (bound at src/Reflex/Patch.hs:104:20)
      apply :: PatchDMap k1 v
               -> PatchTarget (PatchDMap k1 v)
               -> Maybe (PatchTarget (PatchDMap k1 v))
        (bound at src/Reflex/Patch.hs:104:3)
    In the first argument of ‘DMap.mapMaybeWithKey’, namely
      ‘(const $ nothingToJust . getComposeMaybe)’
    In the expression:
      DMap.mapMaybeWithKey (const $ nothingToJust . getComposeMaybe) diff
    In an equation for ‘deletions’:
        deletions
          = DMap.mapMaybeWithKey
              (const $ nothingToJust . getComposeMaybe) diff

src/Reflex/Patch.hs:384:79:
    The second argument of ‘Constant’ should have kind ‘*’,
      but ‘a’ has kind ‘k’
    In the type signature for ‘deleteFunc’:
      deleteFunc :: forall a.
                    k a -> DMapEdit k v a -> Maybe (Constant () a)
    In an equation for ‘apply’:
        apply (PatchDMapWithMove p) old
          = Just $! insertions `DMap.union` (old `DMap.difference` deletions)
          where
              insertions = DMap.mapMaybeWithKey insertFunc p
              insertFunc :: forall a. k a -> DMapEdit k v a -> Maybe (v a)
              insertFunc _
                = \case {
                    DMapEdit_Insert _ v -> Just v
                    DMapEdit_Move _ k -> DMap.lookup k old
                    DMapEdit_Delete _ -> Nothing }
              deletions = DMap.mapMaybeWithKey deleteFunc p
              ....
    In the instance declaration for ‘Patch (PatchDMapWithMove k v)’
cabal: Leaving directory '.'
cabal: Error: some packages failed to install:
reflex-0.5.0 failed during the building phase. The exception was:
ExitFailure 1

**Comment by mpdairy** [comment]
I get the same error as well on 7.10.3. I can compile it with 8.0.1, but then I can't use ghcjsi.

**Comment by ryantrinkle** [comment]
Ah sorry guys!  7.10 support hasn't been at the top of my mind lately.  It looks like we could probably #ifdef these kinds to * rather than being kind-polymorphic.  @hansroland Perhaps you could give that a try?

@mpdairy Hopefully we can get this sorted out soon - but in the mean time, have you tried using regular ghci with reflex-dom?  Everything in reflex-dom itself should work just fine in both, and if you have code that uses JSFFI directly, it's often pretty simple to put in stubs with #ifdef.  I'm not sure how actively people are working on ghcjsi, but it does seem that the bulk of development is happening on ghcjs's 8.0 branch these days.

**Comment by mpdairy** [comment]
Thanks @ryantrinkle, I tried it with regular GHC and it works perfectly! The only change I had to make in the stack yaml (using lts-7.19) was to specify `webkitgtk3-javascriptcore-0.13.2.0` since >=0.14 was missing a WebFrame module. Now I can do ghci and quick compiling checks in emacs, then build with ghcjs as a separate command in the console.

**Comment by hansroland** [comment]
@ryantrinkle I tried with the **develop** branch and unfortunately I got again the *Kind incompatibility* error.
Is *develop* the correct branch to test?

Compiling under 7.10.3 is only necessary, because GHCJS compiled with GHC 8.0.1 does not support *ghci --interactive*. Looking at [http://ghcjs.luite.com](http://ghcjs.luite.com) it seems, that the master version of GHCJS is not yet on GHC 8.

Here again is the error messages:

```
src/Reflex/Patch/DMap.hs:33:45:
    Kind incompatibility when matching types:
      g0 :: k -> *
      Constant () :: * -> *
    Expected type: k1 v1 -> ComposeMaybe v v1 -> Maybe (g0 v1)
      Actual type: k1 v1 -> ComposeMaybe v v1 -> Maybe (Constant () b0)
    Relevant bindings include
      deletions :: DMap k1 g0 (bound at src/Reflex/Patch/DMap.hs:33:11)
      old :: PatchTarget (PatchDMap k1 v)
        (bound at src/Reflex/Patch/DMap.hs:31:26)
      diff :: DMap k1 (ComposeMaybe v)
        (bound at src/Reflex/Patch/DMap.hs:31:20)
      apply :: PatchDMap k1 v
               -> PatchTarget (PatchDMap k1 v)
               -> Maybe (PatchTarget (PatchDMap k1 v))
        (bound at src/Reflex/Patch/DMap.hs:31:3)
```

and 

```
src/Reflex/Patch/DMapWithMove.hs:272:79:
    The second argument of ‘Constant’ should have kind ‘*’,
      but ‘a’ has kind ‘k’
    In the type signature for ‘deleteFunc’:
      deleteFunc :: forall a.
                    k a -> NodeInfo k v a -> Maybe (Constant () a)
    In an equation for ‘apply’:
        apply (PatchDMapWithMove p) old
          = Just $! insertions `DMap.union` (old `DMap.difference` deletions)
          where
              insertions = DMap.mapMaybeWithKey insertFunc p
              insertFunc :: forall a. k a -> NodeInfo k v a -> Maybe (v a)
              insertFunc _ ni
                = case _nodeInfo_from ni of {
                    From_Insert v -> Just v
                    From_Move k -> DMap.lookup k old
                    From_Delete -> Nothing }
              deletions = DMap.mapMaybeWithKey deleteFunc p
              ....
    In the instance declaration for ‘Patch (PatchDMapWithMove k v)’
```

**Comment by lynnard** [comment]
Is there any possible fix for this? I can live with a one-time hot fix on my local copy for now.

**Comment by hansroland** [comment]
Issue fixed with PR #117. Many thanks!
